### PR TITLE
Fix `MegaModel` CI

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Fix `MegaModel` CI (some tests are skipped too). See comments.

### CODE REVIEW

File: src/transformers/models/mega/modeling_mega.py [hunk 1]
```diff
@@ -896,7 +896,7 @@ def softmax_attention(self, query, key, padding_mask, causal_mask):
         # apply causal mask (presumed to be 1/0 for not masked / masked)
         # additive, but convert to 0/-inf (which is not explicitly in the Mega source code)
         if causal_mask is not None:
-            additive_causal_mask = torch.zeros_like(causal_mask, dtype=torch.float)
+            additive_causal_mask = torch.zeros_like(causal_mask, dtype=qk.dtype)
```
Comment (ydshieh) [inline src/transformers/models/mega/modeling_mega.py:899]: fix for `test_generate_fp16`

File: tests/models/mega/test_modeling_mega.py [hunk 2]
```diff
@@ -387,6 +387,8 @@ def check_chunking_shorter_sequence(
         config.use_chunking = True
         config.chunk_size = input_ids.size(1) + 25
         model = MegaModel(config)
+        model.to(torch_device)
+        model.eval()
```
Comment (ydshieh) [inline tests/models/mega/test_modeling_mega.py:391]: device issue, same belows

File: tests/models/mega/test_modeling_mega.py [hunk 3]
```diff
@@ -615,6 +625,39 @@ def test_model_from_pretrained(self):
...
+    def test_cpu_offload(self):
+        super().test_cpu_offload()
+
+    @unittest.skip(reason="Does not work on the tiny model as we keep hitting edge cases.")
+    def test_disk_offload(self):
+        super().test_disk_offload()
+
+    @unittest.skip(reason="Does not work on the tiny model as we keep hitting edge cases.")
+    def test_model_parallelism(self):
+        super().test_model_parallelism()
```
Comment (ydshieh) [inline tests/models/mega/test_modeling_mega.py:638]: Guess I am allowed to skip these.

File: tests/models/mega/test_modeling_mega.py [hunk 4]
```diff
@@ -615,6 +625,39 @@ def test_model_from_pretrained(self):
...
+
+    @unittest.skip(
+        reason=(
+            "Calling `self.attention_function` in `MegaMovingAverageGatedAttention.forward` changes the submodules on "
+            "device 1 to device 0 (also changes `requires_grad`). No idea how this could happen for now."
+        )
+    )
+    def test_multi_gpu_data_parallel_forward(self):
+        super().test_multi_gpu_data_parallel_forward()
```
Comment (ydshieh) [inline tests/models/mega/test_modeling_mega.py:647]: This is super strange, I can't really figure this out - but it's `nn.DataParallel` thing.

File: tests/models/mega/test_modeling_mega.py [hunk 5]
```diff
@@ -615,6 +625,39 @@ def test_model_from_pretrained(self):
...
+    def test_torchscript_simple(self):
+        super().test_torchscript_simple()
+
+    @unittest.skip(reason="Tracing of the dynamically computed `MegaMultiDimensionDampedEma._kernel` doesn't work.")
+    def test_torchscript_output_hidden_state(self):
+        super().test_torchscript_output_hidden_state()
+
+    @unittest.skip(reason="Tracing of the dynamically computed `MegaMultiDimensionDampedEma._kernel` doesn't work.")
+    def test_torchscript_output_attentions(self):
+        super().test_torchscript_output_attentions()
```
Comment (ydshieh) [inline tests/models/mega/test_modeling_mega.py:659]: `self._kernel` of `MegaMultiDimensionDampedEma` is computed dynamically. `torchscript` complains it can't trace a tensor with required grad.

(if input length is changed, it will be re-computed, and the result size will be different - `torchscript` won't allow this anyway)

Comment (ArthurZucker) [reply tests/models/mega/test_modeling_mega.py:659]: We can probably set a max input length and pre-compute, then subsample?

**Review by ArthurZucker — APPROVED** [review]
Thanks for taking care of this

**Review by sgugger — APPROVED** [review]
LGTM, thanks!

File: tests/models/mega/test_modeling_mega.py (hunk 5 above)
Comment (ydshieh) [reply tests/models/mega/test_modeling_mega.py:659]: We can do so, but it still won't work, as the first thing `torchscript` complains is about it can't trace a tensor (assigned to the module, here is `self._kernel`) that requires gradient. Even if we pre-compute and set it, it still won't be able to trace.

Comment (ydshieh) [reply tests/models/mega/test_modeling_mega.py:659]: (in fact, in our case, it's already set when  `model(main_input)` is executed, see below)

```python
model(main_input)
traced_model = torch.jit.trace(model, main_input)
```
